### DISCUSSION TITLE
Emacs - Let's surround

### DISCUSSION THREAD

**Discussion by arialdomartini** [arialdomartini/arialdomartini.github.io#29] [General]
Space for discussion around the few articles on Emacs published on [https://arialdomartini.github.io](https://arialdomartini.github.io):

* [Emacs: how to activate the functionality X for all files of type Y?](https://arialdomartini.github.io/emacs-hooks)
* [Emacs - Let's surroung](https://arialdomartini.github.io/emacs-surround)

**Comment by dandrake** [comment]
I was inspired by this and made an improved version that allows you to repeatedly call the function and it will correctly nest the surrounded region: https://gist.github.com/dandrake/864f642850acaa3534cf5029868d12eb. Take a look!

**Comment by TRSx80** [comment]
At the end of your article, you wrote "There are off-the-shelf packages doing this", can you point me to any?  :)

It seems like I came across something like that years ago and I have been looking for it ever since.  In particular, I seem to recall it being able to highlight a region and cycle through various pairs (e.g., `"` or `'` or `<` or `~` etc.).

**Reply by arialdomartini** [reply]
I'm terrible at providing answers. For some reason I missed your comment. I am really ashamed.

I referred to TempEl and YASnippet. I think I should mention them in the post.

**Comment by TRSx80** [comment]
If it makes you feel any better, I just now saw your reply.  :laughing: 

I don't think those are what I am looking for (but perhaps I misunderstand how they work).  What I am looking for (and swear I saw somewhere once) is something that works like the following.  Say you have this text:

`"some text"`

By highlighting that and pressing some key combination it might cycle through (for example):

```
'some text'
~some text~
[some text]
```

... or whatever other pairs one might define.

**Reply by arialdomartini** [reply]
I could really not find it! I thought it could be [smartparens](https://github.com/Fuco1/smartparens) by @Fuco1, but I could not find a function to cycle like you described.

Now I'm curious: should you finally find it, please remember to drop a note here!

**Reply by TRSx80** [reply]
Yes, I've come really close, but have not stumbled across it again.  Been looking for literally years now(!).

> should you finally find it, please remember to drop a note here!

I will try hard to remember!  :smile:

**Comment by TRSx80** [comment]
Yes, I've come really close, but have not stumbled across it again.  Been looking for literally years now(!).

> should you finally find it, please remember to drop a note here!

I will try hard to remember!  :smile:
